### PR TITLE
bpo-33897: Add a 'force' keyword argument to logging.basicConfig().

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue33897

### CODE REVIEW

File: Doc/library/logging.rst [hunk 1]
```diff
@@ -1131,7 +1131,7 @@ functions.
    if no handlers are defined for the root logger.
 
    This function does nothing if the root logger already has handlers
-   configured for it.
+   configured except when a keyword argument *force* is set as true.
```
Comment (vsajip) [inline Doc/library/logging.rst]: ... has handlers configured, unless the keyword argument *force* is set to ``True``.

File: Doc/library/logging.rst [hunk 2]
```diff
@@ -1183,6 +1183,14 @@ functions.
    |              | with 'filename' or 'stream' - if both are   |
    |              | present, a ``ValueError`` is raised.        |
    +--------------+---------------------------------------------+
+   | ``force``    | If this keyword argument is specified as    |
+   |              | true, basicConfig removes all handlers of   |
```
Comment (vsajip) [inline Doc/library/logging.rst]: ... is specified as true, any existing handlers attached to the root logger are removed and closed, before carrying out the configuration as specified by the other arguments.

File: Lib/logging/__init__.py [hunk 3]
```diff
@@ -1793,7 +1793,8 @@ def basicConfig(**kwargs):
     Do basic configuration for the logging system.
 
     This function does nothing if the root logger already has handlers
-    configured. It is a convenience method intended for use by simple scripts
+    configured except when a keyword argument 'force' is set as true.
```
Comment (vsajip) [inline Lib/logging/__init__.py]: ... has handlers configured, unless the keyword argument *force* is set to ``True``.

File: Lib/logging/__init__.py [hunk 4]
```diff
@@ -1821,13 +1822,18 @@ def basicConfig(**kwargs):
               handlers, which will be added to the root handler. Any handler
               in the list which does not have a formatter assigned will be
               assigned the formatter created in this function.
-
+    force     If this keyword is specified as true, basicConfig removes
```
Comment (vsajip) [inline Lib/logging/__init__.py]: ... is specified as true, any existing handlers attached to the root logger are removed and closed, before carrying out the configuration as specified by the other arguments.

File: Lib/test/test_logging.py [hunk 5]
```diff
@@ -3901,6 +3901,20 @@ def test_handlers(self):
...
+    def test_force(self):
+        with support.captured_stdout() as output:
+            logging.basicConfig(stream=sys.stdout)
+            logging.warning('warn')
+            logging.info('info')
+            self.assertEqual(len(logging.root.handlers), 1)
+            logging.basicConfig(level=logging.INFO, stream=sys.stdout,
+                                force=True)
+            logging.info('info')
+            sys.stdout.seek(0)
```
Comment (vsajip) [inline Lib/test/test_logging.py]: Is this `seek()` call needed? Doesn't `getvalue()` get the entire output, anyway?

File: Misc/NEWS.d/next/Library/2018-06-23-18-09-28.bpo-33897.Hu0yvt.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+Add a 'force' keyword argument for basicConfig().
```
Comment (vsajip) [inline Misc/NEWS.d/next/Library/2018-06-23-18-09-28.bpo-33897.Hu0yvt.rst]: Added a 'force' keyword argument to logging.basicConfig().

**Review by vsajip — CHANGES_REQUESTED** [review]

File: Lib/test/test_logging.py [hunk 7]
```diff
@@ -3901,6 +3901,19 @@ def test_handlers(self):
         self.assertIs(handlers[2].formatter, f)
         self.assertIs(handlers[0].formatter, handlers[1].formatter)
 
+    def test_force(self):
+        with support.captured_stdout() as output:
+            logging.basicConfig(stream=sys.stdout)
+            logging.warning('warn')
+            logging.info('info')
```
Comment (pablogsal) [inline Lib/test/test_logging.py]: Suggestion: Can we log on `debug` on this line? This way is very clear that the `INFO:root:info` comes after calling `logging.basicConfig(level=logging.INFO, stream=sys.stdout, force=True)` and not before.

Comment (vsajip) [reply Lib/test/test_logging.py]: Good point. Perhaps better yet, we could assert that `output.getvalue()` contains only the `WARNING` line after this call, or immediately before the `logging.info()` call which follows this one.

**Comment by corona10** [comment]
@vsajip @pablogsal 
I updated the unit test to detect the change of handler by using StringIO.

File: Lib/test/test_logging.py [hunk 8]
```diff
@@ -3901,6 +3901,27 @@ def test_handlers(self):
         self.assertIs(handlers[2].formatter, f)
         self.assertIs(handlers[0].formatter, handlers[1].formatter)
 
+    def test_force(self):
+        old_string_io = io.StringIO()
+        new_string_io = io.StringIO()
+        old_handlers = [logging.StreamHandler(old_string_io)]
+        new_handlers = [logging.StreamHandler(new_string_io)]
+        logging.basicConfig(level=logging.WARN, handlers=old_handlers)
```
Comment (vsajip) [inline Lib/test/test_logging.py]: Please use `WARNING` here, `WARN` is only around for backward compatibility.

**Comment by corona10** [comment]
@vsajip Thanks updated!

**Pull request merged**